pynvim/msgpack_rpc/event_loop/asyncio.py

```diff
@@ -188,18 +193,33 @@ async def connect_stdout():
 
     @override
     def _connect_child(self, argv: List[str]) -> None:
-        if os.name != 'nt':
-            # see #238, #241
-            self._child_watcher = asyncio.get_child_watcher()
-            self._child_watcher.attach_loop(self._loop)
-
         async def create_subprocess():
             transport: asyncio.SubprocessTransport  # type: ignore
             transport, protocol = await self._loop.subprocess_exec(
                 self._protocol_factory, *argv)
             pid = transport.get_pid()
             debug("child subprocess_exec successful, PID = %s", pid)
 
+            if os.name != 'nt':
+                # see #238, #241
+                pidfd_works = False
+                if PidfdChildWatcher is not None and hasattr(os, "pidfd_open"):
+                    try:
+                        fd = os.pidfd_open(pid)
+                    except Exception:
+                        pass
+                    else:
+                        os.close(fd)
+                        pidfd_works = True
+
+                if pidfd_works:
+                    watcher = PidfdChildWatcher()
+                else:
+                    watcher = ThreadedChildWatcher()
```

Comment:
Just noting: ThreadedChildWatcher requires python 3.8+, PidfdChildWatcher requires python 3.9+. While we might want to drop support for python3.7, we could also simply fall back to `get_child_watcher` if neither exists.

I would suggest extracting these lines into an inner helper function (say `get_child_watcher`) and return either `PidfdChildWatcher`, `ThreadedChildWatcher`, or `asyncio.get_child_watcher()` as a fallback.